mkosi/__init__.py

```diff
@@ -2218,21 +2218,61 @@ def finalize_initrds(context: Context) -> list[Path]:
     return [build_default_initrd(context)]
 
 
+def expand_kernel_specifiers(text: str, kver: str, token: str, roothash: str, boot_count: str) -> str:
+    specifiers = {
+        "&": "&",
+        "e": token,
+        "k": kver,
+        "h": roothash,
+        "c": boot_count
+    }
+
+    def replacer(match: re.Match[str]) -> str:
+        m = match.group("specifier")
+        if specifier := specifiers.get(m):
+            return specifier
+
+        logging.warning(f"Unknown specifier '&{m}' found in {text}, ignoring")
+        return ""
+
+    return re.sub(r"&(?P<specifier>[&a-zA-Z])", replacer, text)
+
+
 def install_type1(
     context: Context,
     kver: str,
     kimg: Path,
     token: str,
     partitions: Sequence[Partition],
 ) -> None:
-    dst = context.root / "boot" / token / kver
-    entry = context.root / f"boot/loader/entries/{token}-{kver}.conf"
+    roothash = ""
+    if roothash_cli := finalize_roothash(partitions):
+        roothash = roothash_cli.partition("=")[2]
+
+    boot_count = ""
+    if (context.root / "etc/kernel/tries").exists():
+        boot_count = (context.root / "etc/kernel/tries").read_text().strip()
+
+    bootloader_entry = context.config.bootloader_entry_format or "&e-&k"
+    bootloader_entry_dir = bootloader_entry.replace("-", "/")
+
+    def expand_specifiers(text: str) -> str:
+        return expand_kernel_specifiers(
+            text,
+            kver=kver,
+            token=token,
+            roothash=roothash,
+            boot_count=boot_count,
+        )
+
+    dst = context.root / "boot" / expand_specifiers(bootloader_entry_dir)
+    entry = context.root / f"boot/loader/entries/{expand_specifiers(bootloader_entry)}.conf"
```

Comment:
I had it like this before https://github.com/systemd/mkosi/pull/2733#discussion_r1621336222 but this would most likely cause it to use subdirectories for kver which have `-` in them, which to my knowledge is kind of common

Reply:
That's true, good point. Could you add a comment explaining that?